graylog2-server/src/main/java/org/graylog2/periodical/UserPermissionMigrationPeriodical.java

```diff
@@ -0,0 +1,167 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This task migrates users' permission sets from pre-1.2 style to the built-in roles: admin and reader.
+ *
+ * It does so by preserving the existing stream and dashboard permissions, which can still be assigned to individual users.
+ * It is recommend to migrate to using roles exclusively, but for now both styles will work.
+ */
+public class UserPermissionMigrationPeriodical extends Periodical {
+    private static final Logger log = LoggerFactory.getLogger(UserPermissionMigrationPeriodical.class);
+
+    private final UserService userService;
+    private final RoleService roleService;
+
+    @Inject
+    public UserPermissionMigrationPeriodical(final UserService userService, final RoleService roleService) {
+        this.userService = userService;
+        this.roleService = roleService;
+    }
+
+    @Override
+    public void doRun() {
+        final List<User> users = userService.loadAll();
+        final String adminRoleId = roleService.getAdminRoleObjectId();
+        final String readerRoleId = roleService.getReaderRoleObjectId();
+
+        for (User user : users) {
+            if (user.isLocalAdmin()) {
+                log.debug("Skipping local admin user.");
+                continue;
+            }
+
+            final Set<String> fixedPermissions = Sets.newHashSet();
+            final Set<String> fixedRoleIds = Sets.newHashSet(user.getRoleIds());
+
+            final HashSet<String> permissionSet = Sets.newHashSet(user.getPermissions());
+
+            boolean hasWildcardPermission = permissionSet.contains("*");
+
+            if (hasWildcardPermission && !user.getRoleIds().contains(adminRoleId)) {
+                // need to add the admin role to this user
+                fixedRoleIds.add(adminRoleId);
+            }
+
+            final Set<String> basePermissions = RestPermissions.readerPermissions(user.getName());
+            final boolean hasCompleteReaderSet = permissionSet.containsAll(basePermissions);
+
+            // only migrate the user if it looks like a pre-1.2 user:
+            //   - it has no roles
+            //   - it has the base reader permission set
+            //   - it has the wildcard permissions
+            if (!user.getRoleIds().isEmpty() && hasCompleteReaderSet && hasWildcardPermission) {
+                log.debug("Not migrating user {}, it has already been migrated.", user.getName());
+                continue;
+            }
+            if (hasCompleteReaderSet && !user.getRoleIds().contains(readerRoleId)) {
+                // need to add the reader role to this user
+                fixedRoleIds.add(readerRoleId);
+            }
+            // filter out the individual permissions to dashboards and streams
+            final ArrayList<String> dashboardStreamPermissions = Lists.newArrayList(
```

Comment:
Minor nitpick: Why declare the variable as `ArrayList` (implementation) instead of `List` (interface)?